Use client entity contract instead of implementation.

saurabhrajguru opened the pull request:
The plugin registrations methods are expecting "ClientEntity" concrete implementation.
This could be problem when consumers are referencing "IMessageSender, IMessageReceiver" cannot register the plugin.

This change updates extensions methods with "IClientEntity" to make extension method available to consumer referencing contract and concrete implementations.

SeanFeldman commented:
:white_check_mark: [Build ServiceBus.AttachmentPlugin 0.0.473-preview completed](https://ci.appveyor.com/project/seanfeldman/servicebus-attachmentplugin/builds/25380725) (commit https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin/commit/e60485c5f0 by @)

SeanFeldman commented:
@saurabhrajguru that's a good change, thank you.

Given this is a public API, a change such as this one would be considered a breaking change.
To avoid breaking change, you could add overloads to support `IClientEntity` in addition to the existing methods accepting `ClientEntity`. I will take care of the consolidation for the next major.

SeanFeldman commented:
:x: [Build ServiceBus.AttachmentPlugin 0.0.474-preview failed](https://ci.appveyor.com/project/seanfeldman/servicebus-attachmentplugin/builds/25408272) (commit https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin/commit/01eb5e2cb3 by @)

saurabhrajguru commented:
@SeanFeldman 
Thank you for the direction.
I have reverted that change and just added the overloads for extensions.
I haven't updated tests yet for the API approvals.
Let me know if anything is missing or required in this.

SeanFeldman commented:
Once you approve and push API changes (API approvals), this should be good to be merged and released 🙂

saurabhrajguru commented:
I just updated the API approvals.
Sorry I was under impression that API approvals were approved by someone. 🤣
Thanks

SeanFeldman commented:
:white_check_mark: [Build ServiceBus.AttachmentPlugin 0.0.475-preview completed](https://ci.appveyor.com/project/seanfeldman/servicebus-attachmentplugin/builds/25437274) (commit https://github.com/SeanFeldman/ServiceBus.AttachmentPlugin/commit/84e353e346 by @)

SeanFeldman commented:
API Approval is intended to fail whenever there are public API changes. That way, it catches breaking changes (as you've seen with this PR earlier) and facilitates the conversation about what needs to be done.
As a contributor, you can update the approved version and commit as a part of the PR (which you're doing now), which also surfaces up during PR review. Either way, the changes are not slipping through and are reviewed to ensure API is not broken and there are no surprises.

Thank you, @saurabhrajguru. I'll merge this and release next week (travelling).

SeanFeldman commented:
@saurabhrajguru what's your twitter handle? I'd like to give you the credit when announcing it. Thank you.

saurabhrajguru commented:
Thank you @SeanFeldman 
My twitter handle is https://twitter.com/saurabhrajguru
